### ISSUE TITLE
Deprecations on all components to disappear or change in Topology Refactor

### ISSUE THREAD

**Issue by dotsdl** [MDAnalysis/mdanalysis#599]
The [0.16](https://github.com/MDAnalysis/mdanalysis/milestones/Topology%20refactor%20-%200.15) release features a complete overhaul of the topology system, giving better performance and consistency in a major part of MDAnalysis' core functionality. This also includes the disappearance of some API elements that no longer made sense; these **must** be marked as deprecated in the 0.15 release.

Add to this list, and check when marked deprecated in `develop` branch.

Deprecate and recommend alternative:
- [x] `AtomGroup.get_*` getters
- [x] `AtomGroup.set_*` setters
- [x] `ResidueGroup.get_*` getters
- [x] `ResidueGroup.set_*` setters
- [x] `SegmentGroup.get_*` getters
- [x] `SegmentGroup.set_*` setters

Deprecate and indicate workaround (eg use RG.atoms.names)
- [x] `Residue` as AtomGroup
- [x] `ResidueGroup` as AtomGroup
- [x] `Segment` as AtomGroup
- [x] `SegmentGroup` as AtomGroup

Indicate when the deprecation will happen
- [x] CamelCase methods (#663)

**Comment by richardjgowers** [comment]
Ahh this is a good point.  We'll have to start a list on the wiki or something... once everything is finalised

**Comment by orbeckst** [comment]
[Wiki/Issue363-Changes](https://github.com/MDAnalysis/mdanalysis/wiki/Issue363-Changes) by @richardjgowers :  
* Is this a second big API break? In particular the fact that we now need explicit `.atom` in order to get at positions is probably going to break code, isn't it.
* Also, is it clear what `ResidueGroup.atoms.positions` (#411) is returning?

**Comment by richardjgowers** [comment]
@orbeckst Yes. If you're using Residues, ResidueGroups Segments or SegmentGroups as AtomGroups, (accessing atom level properties such as positions off them) then yes it will break things.

WRT #411, an XGroup now always returns things of len(XGroup), with the exception being `.atoms` `.residues` and `.segments` which return a different XGroup

`RG.atoms` returns a set* of the Atoms in that ResidueGroup
`RG.atoms.positions` returns all the positions of all the atoms in the RG

 * I think we made all level shifts (`.atoms`, `.residues` and `.segments`) return sets

**Comment by kain88-de** [comment]
@richardjgowers @dotsdl the changes seem to be very nice. For the changes page. Could you maybe create some real-code example how things will change from the current Topology Model to the new Model. I also would'nt mind if you start doing that in a notebook that you upload and link somewhere (a gist or maybe already as a new blog-post).

**Comment by dotsdl** [comment]
>  I think we made all level shifts (`.atoms`, `.residues` and `.segments`) return sets

Correct. Doing

    AG.atoms

will give an `AtomGroup` containing only one instance of each atom represented within `AG` in atom index order. The same goes for:

    RG.atoms
    SG.atoms

The same principle applies for `.residues` and `.segments` from any `Group`. You always get what amounts to a set.

**Comment by richardjgowers** [comment]
I've been thinking about this, and it's a little complicated to add in deprecation warnings, but I think it's possible.  So for example, `RG.masses` currently just inherits the `AG.masses` method, so you can't actually decorate the `RG` version as it doesn't exist as its own object.

But if we made a `DeprecatedAtomGroup` class, we could then deprecate everything...

``` python
class AtomGroup
    # mostly unchanged

class DeprecatedAtomGroup(AtomGroup)
    # add in some magic here to give deprecation warnings
    # maybe even hacking into getattr.......

class Residue(DeprecatedAtomGroup)

class RG(DeprecatedAtomGroup)
```

**Comment by orbeckst** [comment]
On 5 Jan, 2016, at 05:35, Richard Gowers wrote:

> I've been thinking about this, and it's a little complicated to add in deprecation warnings, but I think it's possible. So for example, RG.masses currently just inherits the AG.masses method, so you can't actually decorate the RG version as it doesn't exist as its own object.
> 
> But if we made a DeprecatedAtomGroup class, we could then deprecate everything...
> 

There's still @kain88-de's valid question as to should we deprecate if we cannot provide an alternative yet. 

I think the answer is that this will be a second API break and we need to think about how to mitigate this. The last break already annoyed users, as I was being told.

**Comment by orbeckst** [comment]
Question about 

>  I think we made all level shifts (.atoms, .residues and .segments) return sets

Are these groups unordered or ordered, i.e. like real sets (random order) or lists with unique elements in defined order?

**Comment by richardjgowers** [comment]
I don't think @dotsdl and I really 100% decided on the set idea, it just seemed like the neatest solution.

So considering a trivial system with 4 atoms 2 residues:
```
(a1, r1)
(a2, r1)
(a3, r2)
(a4, r2)
```

I think upshifts should (and currently do?) return a set:

`AtomGroup(a1, a2, a3).residues = RG(r1, r2)`

`AtomGroup(a1, a3, a2).residues = RG(r1, r2)`

It's the downshifts that are a little confusing, so this seems clear

`RG(r1, r2).atoms - AG(a1, a2, a3, a4)`

But then others are more confusing

`RG(r1, r2, r1).atoms = AG(a1, a2, a3, a4) or AG(a1, a2, a3, a4, a1, a2)`

Or more generically:

`RG.atoms = set(R.atoms for R in RG) or [R.atoms for R in RG]`


None of this will require much work to change on the new branch, so a discussion is welcome!

**Comment by orbeckst** [comment]
The current implementation does something like
```python
RG.atoms = set(R.atoms for R in RG)
```

I also wondered if you would want duplicate atoms for `RG(r1, r2, r1).atoms` and I am tending towards the existing behavior that is also semantically more clear: "Give me the atoms contained in these residues". If you want specific atom orderings then you need to use `+` or `select_atoms(sel1, sel2, sel1)`. However: discuss!

**Comment by dotsdl** [comment]
> Are these groups unordered or ordered, i.e. like real sets (random order) or lists with unique elements in defined order?

They are always ordered by `{atom,residue,segment}` index.

**Comment by dotsdl** [comment]
To elaborate further on the *why*...

Making `*Group.atoms`, `*Group.residues`, and `*Group.segments` always yield ordered sets, no matter the level coming from or going to, is the simplest rule that also accommodates an intuitive use. If one wants the resids of each atom in an `AtomGroup`, one can already get that with:

    AG.resids

If one wants the resids of the residues **represented** among the atoms in the `AtomGroup`, one can then do:

    AG.residues.resids

Likewise, if one wants an `AtomGroup` with no duplicates, that's easy with:

   AG.atoms

The point is, making `.atoms`, `.residues`, `.segments` always yield ordered sets is a simple rule that also affords functionality that is gained, but not lost anywhere.

**Comment by dotsdl** [comment]
TLDR: Any other scheme requires probably different rules for what you get depending on which level you are coming from, which gets confusing and unintuitive real fast. We considered not using sets for about a day before we realized other schemes were hard to explain even to ourselves.

**Comment by orbeckst** [comment]
On 5 Jan, 2016, at 12:47, David Dotson wrote:

> Likewise, if one wants an AtomGroup with no duplicates, that's easy with:
> 
> AG.atoms
> 
> The point is, making .atoms, .residues, .segments always yield ordered sets is a simple rule that also affords functionality that is gained, but not lost anywhere.
> 

I am not so sure that I like the fact that you can have `AG != atoms`... at the moment

   ag = u.atoms[[0,1,0]]
   ag.atoms == ag

You might break code in rather subtle ways.

**Comment by orbeckst** [comment]
On 5 Jan, 2016, at 12:50, David Dotson wrote:

> TLDR: Any other scheme requires probably different rules for what you get depending on which level you are coming from, which gets confusing and unintuitive real fast. We considered not using sets for about a day before we realized other schemes were hard to explain even to ourselves.

Fair enough, but I think the discussion here really shows that we have to be careful how we introduce this and how we prepare users. It's all good to be excited about a sleek new implementation (which is great, don't get me wrong!) but you also have a responsibility towards the user base � and most of them hate having to debug code that "used to work". Thus, it should be very clear what changes (and the wiki page is a starting point). I also liked the idea of writing a blog post *before* a merge, something like a vision for what needs to be changed and why and explain some of the rationale.

**Comment by dotsdl** [comment]
> I am not so sure that I like the fact that you can have `AG != atoms`... at the moment

>    ag = u.atoms[[0,1,0]]
>    ag.atoms == ag

> You might break code in rather subtle ways.

What is the point of `ag.atoms` if it gives you exactly the same object as simply `ag`? I get that there's probably a lot of code that uses these semantics, but I think it makes about as much sense as if a numpy array had a `.array` property that just pointed to itself.

**Comment by orbeckst** [comment]
Because the underlying principle was "everything is an AtomGroup", it was very convenient to either use the object, let's call it `g` itself or `g.atoms`, if you were not sure that `g` was actually a base `AtomGroup` and not e.g. a `ResidueGroup`. It made it much more convenient to use groups at different levels of the [hierarchy](http://pythonhosted.org/MDAnalysis/documentation_pages/core/AtomGroup.html?highlight=hierarchy#class-hierarchy).

What your are proposing for `.atoms` changes the meaning quite dramatically, from "these are the atoms as we have them recorded and arranged in this group" to "unique atoms in the group".

I am not saying that we can't change this but one has to be a bit sensitive to current usage.

**Comment by kain88-de** [comment]
How do you say an atom is unique? Do you go with the ID of the atom? 

I still think it would be very nice to have an example where old code has to be changed or would be broken by the changes. That would make it easier to see how the changes would affect users.

**Comment by richardjgowers** [comment]
@kain88-de yeah, an Atom is just an index.  Or rather, all Atoms/Residues/Segments are uniquely defined according to (Universe, Level (A/R/S), Index).  The class defines the first two, so within the class you just worry about indices.

**Comment by dotsdl** [comment]
@orbeckst I think there are a few possible options for this whole scheme we can debate over to help focus discussion

### 1. Everything gives a set
Same rule for everything, in which a call to `.atoms`, `.residues`, `.segments` always gives a group of unique elements represented at that level.

Call                    |  Gives
---------------------   |  ------------------------------------------
`AtomGroup.atoms`       |  ordered `AtomGroup` of unique atoms represented in `AtomGroup`
`AtomGroup.residues`    |  ordered `ResidueGroup` of unique residues represented in `AtomGroup`
`AtomGroup.segments`    |  ordered `SegmentGroup` of unique segments represented in `AtomGroup`
`ResidueGroup.atoms`    |  ordered `AtomGroup` of unique atoms represented in `ResidueGroup`
`ResidueGroup.residues` |  ordered `ResidueGroup` of unique residues represented in `ResidueGroup`
`ResidueGroup.segments` |  ordered `SegmentGroup` of unique segments represented in `ResidueGroup`
`SegmentGroup.atoms`    |  ordered `AtomGroup` of unique atoms represented in `SegmentGroup`
`SegmentGroup.residues` |  ordered `ResidueGroup` of unique residues represented in `SegmentGroup`
`SegmentGroup.segments` |  ordered `SegmentGroup` of unique segments represented in `SegmentGroup`


### 2. Everything gives a set **except** for the same level
Same as **1**, with the exception that at the same level you get back the same object. Calling `ResidueGroup.atoms` gets you a set always, whereas `AtomGroup.atoms` does not.

Call                    |  Gives
---------------------   |  ------------------------------------------
`AtomGroup.atoms`       |  `AtomGroup` identical to the calling `AtomGroup`
`AtomGroup.residues`    |  ordered `ResidueGroup` of unique residues represented in `AtomGroup`
`AtomGroup.segments`    |  ordered `SegmentGroup` of unique segments represented in `AtomGroup`
`ResidueGroup.atoms`    |  ordered `AtomGroup` of unique atoms represented in `ResidueGroup`
`ResidueGroup.residues` |  `ResidueGroup` identical to the calling `ResidueGroup`
`ResidueGroup.segments` |  ordered `SegmentGroup` of unique segments represented in `ResidueGroup`
`SegmentGroup.atoms`    |  ordered `AtomGroup` of unique atoms represented in `SegmentGroup`
`SegmentGroup.residues` |  ordered `ResidueGroup` of unique residues represented in `SegmentGroup`
`SegmentGroup.segments` |  `SegmentGroup` identical to the calling `SegmentGroup`


### 3. Order and repeat elements preserved as much as possible
At the `atom` level, calling `.atoms`, `.residues`, or `.segments` gets a group with an element for each atom. 

At the `residue` level, calling `.atoms` gives an `AtomGroup` with repeats if the calling `ResidueGroup` itself had repeats, and atoms are given in the same order as the residues in the `ResidueGroup`. Calling `.residues` and `.segments` gets a group with an element for each residue.

At the `segment` level, calling `.residues` gives a `ResidueGroup` with repeats if the calling `SegmentGroup` itself had repeats, and residues are given in the same order as the segments in the `SegmentGroup`. Calling `.atoms` gives an `AtomGroup` of the atoms represented by each residue in residue order in each segment in segment order in the calling `SegmentGroup`, including repeats. Calling `segments` gets a group with an element for each segment.

Call                    |  Gives
---------------------   |  ------------------------------------------
`AtomGroup.atoms`       |  `AtomGroup` identical to the calling `AtomGroup`
`AtomGroup.residues`    |  `ResidueGroup` of `len(AtomGroup) giving the residue for each atom
`AtomGroup.segments`    |  `SegmentGroup` of `len(AtomGroup) giving the segment for each atom
`ResidueGroup.atoms`    |  `AtomGroup` giving the atoms for each residue in the calling `ResidueGroup`'s order
`ResidueGroup.residues` |  `ResidueGroup` identical to the calling `ResidueGroup`
`ResidueGroup.segments` |  `SegmentGroup` of `len(ResidueGroup)` giving the segment for each residue
`SegmentGroup.atoms`    |  `AtomGroup` giving the atoms for each segment in the calling `SegmentGroup`'s order
`SegmentGroup.residues` |  `ResidueGroup` giving the residues for each segment in the calling `SegmentGroup`'s order
`SegmentGroup.segments` |  `SegmentGroup` identical to the calling `SegmentGroup`

**Comment by dotsdl** [comment]
I'm an advocate for **1**, although the others can technically work.

**Comment by jbarnoud** [comment]
The main difference between **1** and **2** is that, in **1**, `AtomGroup.atoms` removes duplicates, while **2** keeps them. Am I right?

To what extent are `AtomGroup`s mutable? Indeed, the other difference I see is that, in **1**,  `AtomGroup.atoms` return a copy, it does not in **2**.

I foresee less surprise with **1** as we can say that these properties always return a new object ans that they always deduplicate elements.

I am curious about **3**. Is it possible to get (even not directly), with **1**, what **3** would output?

**Comment by richardjgowers** [comment]
@jbarnoud yes I think you're correct about 1 vs 2

What I don't like in **1** is that `AG.atoms` doesn't explicitly say it's a set, I'd much rather that `AG.atoms == AG` and then add `AG.unique == set(AG)`

I think I'd like to invent **4**, which is identical to **3** except `AG.residues` returns a set (and AG.segments, RG.segments)

**Comment by tylerjereddy** [comment]
I'll concede to just skimming this issue, but do let me know if I need to write any lib2to3 fixers for API changes. Extending ten2eleven to provide warning messages where the deprecation architecture is complicated or philosophically not correct might be an idea.

**Comment by dotsdl** [comment]
@jbarnoud Groups at any level are not mutable. The only data a Group carries are its indices in `._ix` that define its members, and keeping this immutable means the objects are meaningfully hashable and usable directly with things like `set`s.

**Comment by dotsdl** [comment]
@richardjgowers I like **4**, but will have to reflect for a bit before I settle on it. :D

**Comment by orbeckst** [comment]
**4** looks useful to me.

Stupid question: Do we still have `Residue` and `Segment` objects, and if so, what are the rules for them?

**Comment by dotsdl** [comment]
We do have `Residue` and `Segment` objects! For those there is no choice to be had, given that all atoms are members of exactly one residue, and all residues members of exactly one segment. Calling:

    Residue.atoms

gives the atoms that are members of that residue, in atom index order, which by definition gives no repeats. Similar story for `Segment.residues` and therefore also `Segment.atoms`. For upward movement, such as `Atom.residue`, `Residue.segment`, there is only ever a single object to return, not a Group.

The only reason this debate is a thing is because Groups can have repeats a particular order.

**Comment by richardjgowers** [comment]
To expand.. `Residue` isn't an AtomGroup, so..
``` python
r = u.residues[0]

r.positions # DOESNT WORK
#instead
r.atoms.positions
r.position  # Will probably return the CoG of the residue (not yet implemented)

```

So in general, `Residue` and `Segment` only give information about the singular object, rather than the plural contents.  We could add in both, so `.mass` returns the sum of `.masses`, but I think we thought that was confusing.

**Comment by dotsdl** [comment]
> We could add in both, so .mass returns the sum of .masses, but I think we thought that was confusing.

For convenience we'll probably include things like `Residue.segid`, since that's unambiguous, but things like `Residue.names` which refer downward it can get confusing real fast.

**Comment by richardjgowers** [comment]
Ok so it looks like we'll have another release in a few weeks, so we need to make a decision here (on the big table above), I'm still thinking option **4**.  Would be good to get a fair quorum here @MDAnalysis/coredevs then we can try and put in all the deprecation decorators.

**Comment by jbarnoud** [comment]
+1 for option **4**

**Comment by orbeckst** [comment]
+1 for option **4**

**Comment by dotsdl** [comment]
Giving the options a read-through again, is there something to be gained by doing option **3** but with the `unique` method always available at any level (as proposed for **4**)? It seems that with option **4** the only way to get, say, a `SegmentGroup` with a segment for each atom in the same order is to do a loop.

Then getting the result of `AtomGroup.segments` in option **4** could be achieved with `AtomGroup.segments.unique()`.

Is this agreeable, or too kludgey? I kind of like it myself.

**Comment by kain88-de** [comment]
I'm undecided if I like 3 or 4 better. But I'm happy with what the majority decides.

**Comment by dotsdl** [comment]
Added my proposed addition to **3** (`unique` method/property) to the proposal list.

**Comment by jbarnoud** [comment]
I like **4** because I find it has consistency, and the `unique` methods make it clear what to expect. Option **3** with the `unique` keyword may be even better, as it is also consistent but more potent.

So maybe I would go for **3** with the `unique` keyword at every level. I don't mid option **4**, though.

**Comment by richardjgowers** [comment]
I don't see the use of getting a len(AG) list of Residues? `AG.resnames` etc will give properties on a per-atom basis.  In my mind there's an implicit reduction when you go upwards, but if someone can give me a good use case maybe?

**Comment by dotsdl** [comment]
@richardjgowers I don't immediately see it myself, considering that yes, you can get at per-atom residue information from an `AtomGroup` directly. Getting a per-atom `ResidueGroup` seems a bit strange, but if you wanted access to, say, the mass of the residue each atom in an `AtomGroup` is a part of, it would be easy to get that:

```python
ag.residues.masses
```

That's the only one I can think of offhand, but this scheme does bring broad consistency to the idea that `AtomGroup` methods generally give per-atom results, even the properties that give higher order objects. For interactive work these are less useful, but I think for complex analyses the way this works here might be really key.

That said, I'll admit I'm torn too.

**Comment by dotsdl** [comment]
However the same effect as I laid out above could be done with:

```python
u.residues[ag.resindices].masses
```
so we don't really lose out either way. @richardjgowers , since everyone who cares appears to be either for **4** or either **3** or **4**, I'll leave it to you to make the final call.

**Comment by richardjgowers** [comment]
Ok foot down, **4** because it keeps current behaviour the most.

**Comment by dotsdl** [comment]
Sounds good. We'll make it so!

**Comment by dotsdl** [comment]
Quick question while I'm adding deprecations to setters in `AtomGroup`: what are `serials`? More importantly, is this an atom-level, residue-level, or segment-level attribute?

**Comment by richardjgowers** [comment]
Not entirely sure, but something like an Atom index afaik

**Comment by dotsdl** [comment]
Looks like it's just `ids` in our new scheme. They're only used in the `PrimitivePDBParser` and `CRDParser`, and they look like they try to use the atom numbering from the file. This is the role played by `Atom.id`, since `Atom.index` is the 0-based index of the atom within our `Universe.`

I needn't add anything, then. But it does change my deprecation message. :D

**Comment by orbeckst** [comment]
Serials are specific to PDB, they are the running atom number. 

Serials are used in CONECT records. 

--
Oliver Beckstein
email: orbeckst@gmail.com


Am Feb 6, 2016 um 12:33 schrieb David Dotson <notifications@github.com>:

**Comment by dotsdl** [comment]
New poll: thoughts on 1D vs. 2D atom attribute access through `ResidueGroup` and `SegmentGroup` objects. At the moment, we do the 1D case, but the 2D case is worth considering, too. See this [gist](https://gist.github.com/dotsdl/5c526d163be04ca366b0) for what I'm talking about, including performance comparisons.

In short, we don't appear to sacrifice any performance for the 2D case, but do gain some functionality that isn't there in any other form at present. If we choose to go this route, it would also apply to residue attributes accessed from the `SegmentGroup` level.

**Comment by orbeckst** [comment]
@dotsdl – just to be clear, what you're proposing is not related to "option **4**" in what collections the various groups return?

Instead you are proposing to have additional structure in attributes, namely what to do with per-atom attributes in higher order groups. 

1. Is your proposal specific to `names` or should it equally apply to *all `Atom` properties*, including `mass`, `charge` etc? 
2.  Will it also apply to e.g. `resname` going up one level for `SegmentGroup` – would this also return an array of segment-sized lists of resnames?

If the proposal applies to all per-atom attributes then it will collide with the idea to implement aggregate properties (e.g. total mass or charge for higher-level groups or center of mass for position). If it only applies to a subset then that's fine but then we loose orthogonality and the user needs to know lots of specifics.

As far as I can tell we are not really loosing anything because the to-be canonical way to get at *all* the atom information (eg `names`) would be `XGroup.atoms.names` anyway and the proposed `XGroup.names` would be additional functionality with no real performance penalty. Did I understand it correctly.

What is a common use case to want to access per-atom properties in residue-sized chunks?

**Comment by orbeckst** [comment]
What happens if I ask for `SegmentGroup.names`?

**Comment by richardjgowers** [comment]
Doesn't make sense (segments don't have names), so gives you an error.
Segment.atoms.names is the new way

**Comment by dotsdl** [comment]
> @dotsdl – just to be clear, what you're proposing is not related to "option **4**" in what collections the various groups return?

Correct. This is a different question, but one that we have yet to confront and requires a clear choice between at least two alternatives.

> Instead you are proposing to have additional structure in attributes, namely what to do with per-atom attributes in higher order groups.

Yes. In the current 'everything-is-an-AtomGroup' scheme, something like `ResidueGroup.names` gets you back the names of all the atoms in represented in the `ResidueGroup`. Since a `ResidueGroup` in the new scheme is actually not an `AtomGroup` at all, but really treated as a group of residues, should `ResidueGroup.names` give an iterable of length the number of residues it has? I think it should, but this is a deviation from the current behavior.

There are performance benefits to making this how things work, too, as demonstrated in the [gist](https://gist.github.com/dotsdl/5c526d163be04ca366b0).

> Is your proposal specific to `names` or should it equally apply to *all `Atom` properties*, including `mass`, `charge` etc?

It would apply to any attribute accessed through a `ResidueGroup` or `SegmentGroup`. If one does
`ResidueGroup.masses`, one already gets an array of masses for each residue, instead of by atom. We also do the same with charges. What I'm proposing will make other attributes that can't be aggregated into a single value, like `names`, give a result that gives a value for each element in the `*Group`.

> Will it also apply to e.g. `resname` going up one level for `SegmentGroup` – would this also return an array of segment-sized lists of resnames?

Doing `SegmentGroup.resnames` would return a list of 1-D arrays, one for each segment, giving the resname for each residue in the segment.

> If the proposal applies to all per-atom attributes then it will collide with the idea to implement aggregate properties (e.g. total mass or charge for higher-level groups or center of mass for position). If it only applies to a subset then that's fine but then we loose orthogonality and the user needs to know lots of specifics.

Not quite. Since those are functions, not properties, we can make the distinction that they do not give back per-atom results. Something like `ResidueGroup.total_mass()` returns a single value, as one would expect. If one instead wanted the `total_mass` of each `Residue`, they should do `ResidueGroup.masses`. 

This proposal just makes it so that getting at topology attributes, for any level, returns an iterable with as many elements as the group they were obtained from. For levels below the level being called from, this would most often be a list of 1-D numpy arrays.

> As far as I can tell we are not really loosing anything because the to-be canonical way to get at *all* the atom information (eg `names`) would be `XGroup.atoms.names` anyway and the proposed `XGroup.names` would be additional functionality with no real performance penalty. Did I understand it correctly.

Correct. Nothing is lost. This is a clear change from current behavior, though.

> What is a common use case to want to access per-atom properties in residue-sized chunks?

Anyone find themselves doing `[residue.indices for residue in rg]`? Doing `rg.indices` would give the same result, but nearly 8 times faster under the new proposal.

**Comment by orbeckst** [comment]
>> What happens if I ask for SegmentGroup.names?
> Doesn't make sense (segments don't have names), so gives you an error. Segment.atoms.names is the new way

I understand that if I want the names of all the atoms in the segments I have to do the 363-canonical `SG.atoms.names`. I am perfectly comfortable with this solution because it's semantically clear and does not require me to know what `*group` I', working with (because every group has `.atoms`). 

My understanding of this discussion was that other groups such as `SegmentGroup` would actually gain access to atom-level attributes. Or would you only allow attributes that belong to the lower level, i.e. in the case of `SegmentGroup` only attributes of `Residue` objects?

**Comment by dotsdl** [comment]
@richardjgowers misspoke. `SegmentGroup`s have a `names` property, although `Segment`s do not. We did this because something like `Segment.masses`, although fundamentally an atom attribute, is treated as an attribute of residues and segments as well. A `Segment` has a mass, so what does `Segment.masses` mean?

> My understanding of this discussion was that other groups such as SegmentGroup would actually gain access to atom-level attributes. Or would you only allow attributes that belong to the lower level, i.e. in the case of SegmentGroup only attributes of Residue objects?

Although originally we considered making level attributes appear only at the level they correspond to, we do have `SegmentGroup.names`, for example, which accesses an atom-level attribute ('names') via collection of segments. We did this for all attributes that go downward, no matter how many levels down. The question at hand is how should these downward attribute accessors generally behave.

One option, of course, is to remove them entirely. But that would be an even bigger departure from current functionality, which features these kinds of accessors.

**Comment by dotsdl** [comment]
Any other opinions on this last question? I'm in favor of moving to the model where topology properties are generally returned as iterables of length equal to the calling object, as opposed to the current behavior.

**Comment by orbeckst** [comment]
Although your proposal goes against the Python Zen of "flat is better than nested" and it changes existing behavior (we always had flat lists of all atoms as for the underlying `AtomGroup`)  I do like that (1) you get as many objects as the original sequence and (2) it is additional functionality.

## Open Question: atom properties at `Segment` level?
I still don't know what `Segment.names` and `SegmentGroup.names` will return.

## Ambiguous attributes: Refer to next lower tier
How are plural attributes like `masses` handled? Will `Segment.masses` refer to `Segment.residues.masses` or `Segment.atoms.masses`? 

To re-inforce the idea of a hierarchy of organizational elements I favor `Segment.masses == Segment.residues.masses`. I don't like leaving it undefined – we might as well make a decision and stick to it. Returning a list of length `len(Segments.residues)` will also be more in-line with "topology properties are generally returned as iterables of length equal to the calling object". Thus I propose to resolve any ambiguous cases by **referring to properties in the next lower tier in the hierarchy**.

## Preliminary outline of the philosophy
So if we were to go ahead with @dotsdl 's proposal we  have to tell users (also taking option **4** onboard):

1. **atoms**:
    - If you want a list of properties corresponding to each individual Atom, use `group.atoms.prop`
2. **residues**: If you want a list of 
     - *residue-level properties* (e.g. resname) then use `group.residues.residueprop`.
     - *atom-level properties* (e.g. name)
        - as a list per atom: `group.atoms.atomprop` (i.e., **1**)
        - as a nested list, one sublist per residue: `group.residues.atomprop`
3. **segments**: If you want a list of 
     - *segment-level properties* (e.g. segname) at segment level: `group.segments.segmentprop`
     - *residue-level properties* (e.g. resname) 
         - as a list per residue: `group.residues.residueprop`
         - as a nested list, one sublist per segment: `group.segments.residueprop`
     - *atom-level properties* (e.g. name) 
        - as a list per atom: `group.atoms.atomprop` (as above)
        - as a nested list, one sublist per segment (??): `group.segments.atomprop` — **IS THIS TRUE?**

**Comment by hainm** [comment]
I suggest to design like mdtraj's Topology. It is simple to use.

**Comment by hainm** [comment]
And what does segment do? Why not only atoms and residues?

**Comment by orbeckst** [comment]
On 12 Feb, 2016, at 17:31, Hai Nguyen wrote:

> And what does segment do? Why not only atoms and residues?

Larger units, such as, say actual molecules.

Atoms are real, molecules are real, residues are convenient ;-)

(But of course, the whole solvent or the membrane could also conveniently be grouped as a segment, so that's why we are not calling it a molecule... that's what fragments are for, but they don't fit into the hierarchy.)

**Comment by dotsdl** [comment]
@hainm our new topology system is more performant, both in speed and memory usage, and more flexible (no hardcodes of attributes) than either the current MDAnalysis topology system or that of mdtraj. There is no reason to regress on this front. The new system also preserves much of the existing intuitive API of MDAnalysis that works quite well, but internally eliminates the possibility of staleness along with the performance benefits.

The only things that need deciding are what attribute access down the hierarchy returns. See the conversation earlier and my gist on the question.

**Comment by dotsdl** [comment]
@orbeckst 

So to sidestep the ambiguity of accessing lower-level attributes from a single object (`Segment`, `Residue`), we've so far left out properties for these. That means that there is no `Segment.names` or `Residue.names`, even though there is a `SegmentGroup.names` and `ResidueGroup.names`. 

Instead, accessing `names` from a `Segment` can be done with either `Segment.atoms.names` or `Segment.residues.names`. Accessing `names` from a `Segment` in this way requires one to be explicit in the level desired, which under my proposal would yield iterables structured according to the level used.

Is this a satisfying resolution? In short, we leave out lower-level properties from `Residue` and `Segment`, requiring explicit level specification. This is internally consistent with the idea that what is returned is an iterable of equal length to the calling object, since `Residue` and `Segment` aren't iterables at all.

**Comment by dotsdl** [comment]
Ah, should have re-read your summary:

> Thus I propose to resolve any ambiguous cases by **referring to properties in the next lower tier in the hierarchy**.

You're in favor of making `Segment` behave roughly like `ResidueGroup` and `Residue` behave roughly like `AtomGroup`, then? I think that's my reading of it. I think it's doable, but would love to hear what @richardjgowers thinks.

**Comment by hainm** [comment]
> @hainm our new topology system is more performant, both in speed and memory usage, and more flexible (no hardcodes of attributes) than either the current MDAnalysis topology system or that of mdtraj. There is no reason to regress on this front. The new system also preserves much of the existing intuitive API of MDAnalysis that works quite well, but internally eliminates the possibility of staleness along with the performance benefits.

yeah, I only make comment based on what I look at MDA and mdtraj tutorials. So I don't know what users want to use. Can you propose a detail example about new topology? Whether the new design will be widely used from users? How often it is used? I am afraid that we are trying to introduce many idea but ending no one ever uses (I throwed away 1/2 pytraj code that I wrote for 1/2 year to get simplicity :D).

**Comment by hainm** [comment]
For example: how often people use '_setter' method for `SegmentGroup`? If this design is for Topology editit, should let user to use another more-dedicated program?

**Comment by dotsdl** [comment]
> Can you propose a detail example about new topology?

The new topology system works under the hood. Most users won't notice it, save for the performance gains and the API breaks here and there (which we are discussing now, in this very thread, and deprecating ahead of time what we can).

See [the wiki](https://github.com/MDAnalysis/mdanalysis/wiki/Issue363-Changes#search_word=363) for some details on what's changed and why. More will be written about it as we get closer to the merge of [`issue-363`](https://github.com/MDAnalysis/mdanalysis/tree/issue-363) into `develop`. You can give it a spin if you like, too. Most things work already.

**Comment by dotsdl** [comment]
> For example: how often people use '_setter' method for SegmentGroup? If this design is for Topology editit, should let user to use another more-dedicated program?

Huh? I have no idea what you're asking. At any rate, this thread is pretty focused on a particular question, and I'd rather we keep it focused on that question until it's resolved. You are welcome to have a look at the `issue-363` branch yourself to see what's going on there.

**Comment by hainm** [comment]
thanks.

sorry to dilute this issue.

**Comment by richardjgowers** [comment]
Hey

I think the nested returns are new since we last spoke, but I think they make sense. The only problem I see with them is that some atom properties don't work like that (mass and charge)  and so it's not completely smooth. Ie RG.Atomprop isnt always always the same shape (but is always the same len or shape[0])

**Comment by orbeckst** [comment]
@dotsdl ,

> You're in favor of making Segment behave roughly like ResidueGroup and Residue behave roughly like AtomGroup, then? I think that's my reading of it. I think it's doable, but would love to hear what @richardjgowers thinks.

Yes, together with your proposal to use nested lists (or iterables) where appropriate.

It is true that this breaks orthogonality (as @richardjgowers pointed out) but I think we can gain more in functionality. At a minimum we do guarantee that for plural attributes one always gets an iterable of `len(group)`, with the items either being aggregated (charge, mass, ...) or being iterables themselves.

**Comment by dotsdl** [comment]
@orbeckst given [your comment here](https://github.com/MDAnalysis/mdanalysis/issues/703#issuecomment-190547898), I take it this means we're fine with leaving out things like `Residue.names` that go downward in the hierarchy? That is, getting an atom-level property from a residue means doing e.g. `Residue.atoms.names`? If so, I'm in favor and will move forward with this (actually, I think we're already there...).

**Comment by orbeckst** [comment]
Yes. 

--
Oliver Beckstein
email: orbeckst@gmail.com


Am Mar 10, 2016 um 19:18 schrieb David Dotson <notifications@github.com>:

**Comment by dotsdl** [comment]
Excellent. There's a fairly clear path forward then. Will keep hammering to get everything in the right place. As I said, I think we're at least 90% there already.

**Comment by mnmelo** [comment]
I'm also cool with that, but what is then the recommended way to have a list of all residue names? (Unduplicated, one name per residue)

**Comment by dotsdl** [comment]
@mnmelo no matter what group you're using to start with, you'd do `grp.residues.resnames`.

**Comment by mnmelo** [comment]
Ok, suspected as much. Cool with me.

**Comment by richardjgowers** [comment]
@dotsdl if you had a RG, doesn't `RG.residues` just return itself?  I think `grp.residues.unique` is the most foolproof way?

**Comment by dotsdl** [comment]
@richardjgowers ah yeah, fair enough. doing `grp.residues` will give you a `ResidueGroup` formed according to the [table](https://github.com/MDAnalysis/mdanalysis/issues/599#issuecomment-169163441) (**4**). 

@mnmelo if you want resnames for each residue but with absolutely no duplicate residues, you would do `grp.residues.unique.resnames` to get this.

**Comment by dotsdl** [comment]
Once #698 is merged, I'll fill in the checklist and we'll close. :D
